### PR TITLE
Write Frameworks and Preprocessors lesson for new HTML/CSS curriculum

### PR DESCRIPTION
**Pull request by msespos**
- [x] You have read [The Odin Projects Contributing Guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md).
 - [x] The title summarizes the change and where it happened, for example: "Fixes punctuation in Clean Code lesson".
 - [x] If the PR is related to an open issue, use a [relevant keyword](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) and reference it with the `#` sign and the issue number.
- [x] If changes are requested, please make the changes in a timely manner. After you submit the changes, request another review from the maintainer (top right).

#### 1.Describe the changes made and include why they are necessary or important:

Create Frameworks and Preprocessors lesson for new HTML/CSS curriculum (largely taken from two preexisting lessons)

#### 2. Related Issue

[TOP-meta issue: Lesson: HTML/CSS - Frameworks and Preprocessors #33](https://github.com/TheOdinProject/top-meta/issues/33)

#33

### CODE REVIEW

**Comment by codyloyd** [comment]
waaaaaaaaaaay too many and too specific learning outcomes... lets keep it under 5, and make them very unspecific.

like.. "By the end of this lesson you should have a general idea what a CSS framework is."
"You will know what some of the most popular frameworks are"

etc.

**Comment by codyloyd** [comment]
knowledge checks should link to things that are presented in this lesson and not completely new sites/links that students haven't seen yet.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+- What is a preprocessor?
+- What are the advantages of using a preprocessor?
+- What does it mean that a preprocessor "compiles to CSS"?
+- What is SASS/SCSS?
+- How can you use SASS to nest CSS?
+- How can you use SASS to work with variables in your code?
+- What are the advantages of using mix-ins with SASS?
+- If you're not using Rails, how do you set up SASS?
+- What is LESS and how is it different from SASS?
+- What other preprocessors are out there?
```
Comment (I3uckwheat) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:25]: For the purposes of TOP, I really don't think this is compatible with our idea of "use vanilla CSS until you are REALLY good at it." At this point, I don't have confidence the people that would reach for these will be at the level they should be to reach for them.

Having awareness of *what* a preprocessor and a framework is, could be important, but it's not something we want the learners to be doing a deep dive on.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+* [In your own words, describe what a CSS framework does.](https://www.browserstack.com/guide/top-css-frameworks)
+* [What are some popular CSS frameworks?](https://dev.to/thenerdydev/best-css-frameworks-in-2021-1f3p)
+* [What are some advantages of using a CSS framework?](https://speckyboy.com/discussing-the-pros-and-cons-of-using-a-css-framework/)
+* [What are some disadvantages of using a CSS framework?](https://speckyboy.com/discussing-the-pros-and-cons-of-using-a-css-framework/)
+* [When might it be better to use vanilla CSS over a framework?](https://dev.to/shubambhasin/vanilla-css-vs-css-frameworks-3ol7)
+* [In your own words, describe what a CSS preprocessor does.](https://developer.mozilla.org/en-US/docs/Glossary/CSS_preprocessor)
+* [What are some popular CSS preprocessors?](https://developer.mozilla.org/en-US/docs/Glossary/CSS_preprocessor)
+* [What are some of the useful features of CSS preprocessor?](https://www.freecodecamp.org/news/css-preprocessors/)
+* [What are some disadvantages of using a CSS preprocessor?](https://adamsilver.io/blog/the-disadvantages-of-css-preprocessors/)
```
Comment (I3uckwheat) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:91]: Some of these items aren't linked in the lesson itself. Please refresh on the guidelines here: https://github.com/TheOdinProject/curriculum/wiki/Learning-Outcomes-vs-Knowledge-Checks

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+### Additional Resources
+
+* [Bootstrap documentation](https://bootstrapdocs.com/v3.0.3/docs/getting-started/)
+* [Foundation documentation](https://get.foundation/sites/docs/)
+* [Tailwind documentation](https://tailwindcss.com/docs)
+* [Bulma documentation](https://bulma.io/documentation/)
+* [SASS documentation](https://sass-lang.com/documentation)
+* [LESS documentation](https://lesscss.org/)
+* [Stylus documentation](https://stylus-lang.com/)
```
Comment (I3uckwheat) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:76]: Having every piece of documentation for each framework/preprocessor seems to be a bit much. The goal of this lesson should be a conceptual overview, the additional resources should also reflect that. I don't think we want to suggest any of these for use.

**Review by I3uckwheat — CHANGES_REQUESTED** [review]
There is a lot of content here, ultimately, I don't think we want to expose learners to TOO much in this realm. It's overwhelming, exposure to the ideas and concepts are important, but drilling into the specific frameworks seems to be too much for our goals. 

I also believe there should be a heavier emphasis on NOT using these during TOP.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+Additionally, the process of overriding a framework's styling or debugging style issues on your page becomes very difficult if you haven't really mastered CSS fundamentals. It is imperative to understand what a framework is doing "under the hood" so you are equipped to handle these issues later (and trust us, you will).
+
+For the reasons above, **we recommend that you continue to focus on the fundamentals of CSS as you continue through the curriculum.** This lesson is merely designed to teach you _about_ frameworks and their role.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:39]: this is a little heavy-hitting at this point lol... I'd go with a more neutral statement.

something like:

> Frameworks can help you get up and running quickly but can constrain you in the long run, and once you've started a project using one, it can be difficult to remove it.  Whether you use them for projects in the future is up to you!

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 1 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:25]: OK, these were actually lifted verbatim from the existing lesson on preprocessors. I will remove them no problem.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 2 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:91]: Will review the guidelines and update accordingly. My mistake.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 3 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:76]: OK, I can just put documentation for like Bootstrap, Foundation and SASS maybe. Sound good?

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+- How can you use SASS to nest CSS?
+- How can you use SASS to work with variables in your code?
+- What are the advantages of using mix-ins with SASS?
+- If you're not using Rails, how do you set up SASS?
+- What is LESS and how is it different from SASS?
+- What other preprocessors are out there?
+
+### Frameworks Overview
+
+Frameworks like [Bootstrap]](https://getbootstrap.com/) and [Foundation](https://get.foundation) do a lot of the heavy lifting of packaging up commonly used CSS code, even icons and interactions (like menu dropdowns), for you. They are designed to abstract away the process of coding intuitive, reusable, and responsive elements. They do this by creating these elements for you, and giving you the tools to inject styles into your HTML elements directly through class name keywords. All you have to do is understand how a framework expects you to lay out your site and which classes it uses to designate its particular batch of styles.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:29]: > They are designed to abstract away the process of coding intuitive, reusable, and responsive elements. They do this by creating these elements for you, and giving you the tools to inject styles into your HTML elements directly through class name keywords.

Let's be more specific about what frameworks do here... so instead of saying "they give you the tools" be more direct.

something like this:
> When it comes down to it a CSS framework is just a bundle of CSS that you can use and access using the classes defined by the framework.  For example, many frameworks provide a class called `.btn` that will add all the needed styles to your buttons, without you having to write any CSS.

**Comment by msespos** [comment]
Thanks for your comments guys. This was my first lesson so I appreciate all of the criticism.

I replied to your comments directly, Briggs. Cody, I'll address the issues you brought up as well. Expect a new version soon.

Thanks for the review!

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+While the two frameworks mentioned above are heavy hitters that have been around for a while, you should be aware that there are quite a few others at this point, as evidenced by all of the lists of popular frameworks on the web. Two other extremely popular frameworks are [Tailwind](https://tailwindcss.com/) and [Bulma](https://bulma.io/). There are many more out there.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:31]: i wouldn't call foundation a heavy hitter anymore lol.   THis doesn't matter too much.. but instead of trying to rank popularity I'd say something like 

> "there are a ton of frameworks for you to choose from, in addition to the ones mentioned above......."

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
+### Introduction
+
+At this point, you have written quite a bit of vanilla HTML and CSS, and learned many design techniques that you will continue to use as you grow as a developer. However, through your experiences, you have no doubt encountered a CSS framework or two. You may have also learned about preprocessors (aka precompilers). Both of these types of tools can make writing CSS more streamlined and less tedious.
+
+You should be aware that at this point in your learning, both frameworks and preprocessors are options for you, but neither of them are necessary. In particular, you'll read a strong argument below AGAINST using frameworks at this point.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:5]: in a later comment I suggest removing the strong "DO NOT USE" statement... so i'd reword this a bit.

maybe say something about trade-offs etc.

also.. the real reason students need to know this stuff is that they're pretty commonly found in workplaces.  Many job requirements list bootstrap or SASS or something.. so it's helpful for students to at least know what they are and where to look once they've determined they need to learn them.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+### Preprocessors Overview
+
+Preprocessors (aka precompilers) are languages that help you write CSS more easily. They  eliminate code repetition and provide all sorts of time-saving and code-saving features. Some of the standard preprocessors in use are [SASS](https://sass-lang.com/), [LESS](https://lesscss.org/)and [Stylus](https://stylus-lang.com/).
+
+Note: If you've been using Rails, you'll be familiar with SASS already since it's baked in.  If you're not using Rails, you may need to add some additional steps to your workflow to get it set up, but it's usually worth it.
+
+Some of the useful features of preprocessors include:
+
+1. Programming in CSS but with more options available to you.
+2. Including variables in your CSS (not an option in vanilla CSS).
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:50]: this is false

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+3. Including loops in your CSS (again not an option in vanilla CSS).
+4. Including if/else statement in your CSS (see the pattern here?).
+5. Allowing you to create reusable code.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:53]: this is available to regular css

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+### Preprocessors Overview
+
+Preprocessors (aka precompilers) are languages that help you write CSS more easily. They  eliminate code repetition and provide all sorts of time-saving and code-saving features. Some of the standard preprocessors in use are [SASS](https://sass-lang.com/), [LESS](https://lesscss.org/)and [Stylus](https://stylus-lang.com/).
+
+Note: If you've been using Rails, you'll be familiar with SASS already since it's baked in.  If you're not using Rails, you may need to add some additional steps to your workflow to get it set up, but it's usually worth it.
+
+Some of the useful features of preprocessors include:
+
+1. Programming in CSS but with more options available to you.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:49]: "programming in css but with more options" is kind of vague... i don't actually know what you're referring to.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+2. Including variables in your CSS (not an option in vanilla CSS).
+3. Including loops in your CSS (again not an option in vanilla CSS).
+4. Including if/else statement in your CSS (see the pattern here?).
+5. Allowing you to create reusable code.
+6. And a lot more features, depending on which preprocessor you use.
+
+As you can see, preprocessors are extremely helpful tools; however, there are some potential disadvantages to using them. See the reading below for one list of disadvantages.
```
Comment (codyloyd) [inline html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:56]: So here i would say something like:

> preprocessors _do_ have some unique and helpful tools, but many of their most helpful features have been implemented in vanilla CSS, so it might not be worth the overhead of learning one unless you think you really _need_ these features.

**Review by codyloyd — CHANGES_REQUESTED** [review]

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 4 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:29]: So I agree with you; this is lifted from an existing lesson and I actually tried to tone it down. Will replace with your suggestion.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 5 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: OK will do.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 6 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:21]: OK

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 7 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md:5]: Got it

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 1 above)
Comment (codyloyd) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: ah that makes sense... we've been kind of redefining what belongs in the "learning outcomes" section of our lessons.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 8 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: My bad; sorry about this. I didn't realize. I think I got some outdated info.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 3 above)
Comment (codyloyd) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: Bootstrap, Tailwind and SASS imo.  Foundation isn't cool anymore

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 9 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: OK

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 10 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: I think I'm going to remove the "useful features" section entirely.

File: html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md (hunk 11 above)
Comment (msespos) [reply html_css/v2/main-course/intermediate_css/frameworks-preprocessors.md]: Will use that.

**Comment by msespos** [comment]
Thanks for all of your comments guys. Will put another draft up in a bit.
